internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -903,31 +969,177 @@ func (r *AgentServiceConfigReconciler) newImageServiceDeployment(ctx context.Con
-			deployment.Spec.Template.Spec.NodeSelector = r.NodeSelector
+			statefulSet.Spec.Template.Spec.NodeSelector = r.NodeSelector
 		} else {
-			deployment.Spec.Template.Spec.NodeSelector = map[string]string{}
+			statefulSet.Spec.Template.Spec.NodeSelector = map[string]string{}
 		}
 
 		if r.Tolerations != nil {
-			deployment.Spec.Template.Spec.Tolerations = r.Tolerations
+			statefulSet.Spec.Template.Spec.Tolerations = r.Tolerations
 		} else {
-			deployment.Spec.Template.Spec.Tolerations = []corev1.Toleration{}
+			statefulSet.Spec.Template.Spec.Tolerations = []corev1.Toleration{}
 		}
 		return nil
 	}
 
-	return deployment, mutateFn, nil
+	return statefulSet, mutateFn, nil
+}
+
+func (r *AgentServiceConfigReconciler) reconcileImageServiceStatefulSet(ctx context.Context, log logrus.FieldLogger, instance *aiv1beta1.AgentServiceConfig) error {
+	statefulSet, mutateFn, err := r.newImageServiceStatefulSet(ctx, log, instance)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		// delete old deployment if it exists and we've created the new stateful set correctly
+		// TODO: this can be removed when we no longer support upgrading from a release that used deployments for the image service
+		// NOTE: this relies on the err local variable being set correctly when the function returns
+		if err == nil {
+			_ = r.Client.Delete(ctx, &appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      imageServiceName,
+					Namespace: r.Namespace,
+				},
+			})
+		}
+	}()
+
+	key := client.ObjectKeyFromObject(statefulSet)
+	if err = r.Client.Get(ctx, key, statefulSet); err != nil {
+		// if the statefulset doesn't exist, create it
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		log.Info("Creating initial image service stateful set")
+		if err = mutateFn(); err != nil {
+			return err
+		}
+		if err = r.Client.Create(ctx, statefulSet); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	existing := statefulSet.DeepCopyObject().(*appsv1.StatefulSet)
+	if err = mutateFn(); err != nil {
+		return err
+	}
+
+	if equality.Semantic.DeepEqual(existing, statefulSet) {
+		// no update needed
+		return nil
+	}
+
+	if equality.Semantic.DeepEqual(existing.Spec.VolumeClaimTemplates, statefulSet.Spec.VolumeClaimTemplates) {
+		log.Info("Updating image service statful set in-place")
+		// if we're updating something other than the volumes, just do a regular update
+		if err = r.Client.Update(ctx, statefulSet); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	log.Info("Recreating image service stateful set on volume claim template update")
+
+	var pvcs []*corev1.PersistentVolumeClaim
+	pvcs, err = r.pvcsForStatefulSet(ctx, existing)
+	if err != nil {
+		return err
+	}
+
+	// need to delete and re-create statefulset because the volumes have changed
+	if err = r.Client.Delete(ctx, existing); err != nil {
+		return err
+	}
+
+	for _, pvc := range pvcs {
+		if err = r.Client.Delete(ctx, pvc); err != nil {
+			return err
```

Comment:
If there is an error here, but the StatefulSet deletion was successful, will the code ever return here and re-try deleting these PVCs?

Reply:
No, I don't think it will.
I suppose we'd have to reconcile the resulting pvcs against the image storage spec directly.
But even then we'd still have to delete the stateful set to get them properly recreated.

Maybe the condition for recreate vs update should be "the statefulset claim templates or any of the pvcs are different from the image storage spec"? Then we would recreate the statefulset and pvcs if either are out of sync.

Do you think there's something more targeted we could do?

Reply:
We could:
* put a finalizer on the StatefulSet
* at the top of reconciling the StatefulSet, check the deletiontimestamp (as mentioned above), and when it's set, go into cleanup mode.

Cleanup mode:
* check if PVCs still exist
* if so, delete them and return
* if not, remove the finalizer